File: solutions/python/bfs/shortest_path.py
from collections import deque, defaultdict

NO_PATH = -1

def build_graph(edges):
  graph = defaultdict(list)
  
  for node_one, node_two in edges:
    graph[node_one].append(node_two)
    graph[node_two].append(node_one)
    
  return graph

def shortest_path(edges, start_node, end_node):
  graph = build_graph(edges)
  
  start_distance = 0
  
  queue = deque([(start_node, start_distance)])
  visited = set([start_node])
  
  while queue:
    node, distance = queue.popleft()
    
    if node == end_node:
      return distance
    
    for neighbor in graph[node]:
      if neighbor in visited:
        continue
      visited.add(neighbor)
      
      queue.append((neighbor, distance + 1))
        
  return NO_PATH

File: solutions/javascript/bfs/shortestPath.js
/*
Given an array of edges of an undirected graph and two nodes, return the length of the shortest path between these two nodes. If no such path exists, return -1. Note that the length of a path is the number of edges in the path, not the number of nodes.
*/

const NO_PATH_FOUND = -1;

const buildGraph = (edges) => {
    const graph = {};

    for (const edge of edges) {
        const [nodeOne, nodeTwo] = edge;

        const nodeOneInGraph = graph.hasOwnProperty(nodeOne);
        const nodeTwoInGraph = graph.hasOwnProperty(nodeTwo);

        if (!nodeOneInGraph) graph[nodeOne] = [];
        if (!nodeTwoInGraph) graph[nodeTwo] = [];

        graph[nodeOne].push(nodeTwo);
        graph[nodeTwo].push(nodeOne);
    }

    return graph;
};

const getShortestPathLength = (edges, startNode, targetNode) => {
    const graph = buildGraph(edges);

    const queue = [];
    const visited = new Set();

    queue.push({
        node: startNode,
        distanceFromStart: 0,
    });
    visited.add(startNode);

    while (queue.length > 0) {
        const { node, distanceFromStart } = queue.shift();

        if (node === targetNode) return distanceFromStart;

        const neighbors = graph[node];
        for (const neighbor of neighbors) {
            if (visited.has(neighbor)) continue;
            visited.add(neighbor);

            queue.push({
                node: neighbor,
                distanceFromStart: distanceFromStart + 1,
            });
        }
    }

    return NO_PATH_FOUND;
};

const edges = [
    ['w', 'x'],
    ['x', 'y'],
    ['z', 'y'],
    ['z', 'v'],
    ['w', 'v'],
];
const startNode = 'w';
const targetNode = 'z';

console.log(`Your answer: ${getShortestPathLength(edges, startNode, targetNode)}`);
console.log(`Correct answer: ${2}`);
console.log();
